.github/workflows/ci.yml

```diff
@@ -1,39 +1,56 @@
-# Build gen-smtp on various OTP versions
-
 name: CI
-on:
-  push:
-    branches:
-      - master
-  pull_request:
-    branches:
-      - master
+
+on: [push, pull_request]
+
 jobs:
+
   ci:
-    name: Checks and tests on OTP-${{matrix.otp_vsn}}, ${{matrix.os}}, ${{matrix.profile}}
-    runs-on: ${{matrix.os}}
+    name: Test on OTP ${{ matrix.otp }} / Profile ${{ matrix.profile }}
+    runs-on: ${{ matrix.os }}
 
     strategy:
       matrix:
-        otp_vsn: [21.3, 22.3, 23.3, 24.1]
         os: [ubuntu-latest]
-        profile: [test, ranch18, ranch20]
-        exclude:
-            - otp_vsn: 24.1     # ranch 1.7 does not work on OTP24+
-              profile: test
-
-    container:
-      image: erlang:${{matrix.otp_vsn}}
+        otp: ["24", "23", "22"]
+        rebar3: ["3.17.0"]
+        profile: [test, ranch_v2]
+        include:
+          - os: ubuntu-latest
+            otp: "21"
+            rebar3: "3.15.2"
+            profile: "test"
+          - os: ubuntu-latest
+            otp: "21"
+            rebar3: "3.15.2"
+            profile: "ranch_v2"
 
     steps:
       - uses: actions/checkout@v2
-      - name: xref
+
+      - uses: erlef/setup-beam@v1
+        with:
+          otp-version: ${{ matrix.otp }}
+          rebar3-version: ${{ matrix.rebar3 }}
+
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/rebar3
+            _build
+          key: ${{ runner.os }}-erlang-${{ matrix.otp }}-${{ hashFiles('**/rebar.lock') }}
+          restore-keys: ${{ runner.os }}-erlang-${{ matrix.otp }}-
```

Comment:
I never used `acgions/cache` before... Is it safe to cache `_build` directory?

Reply:
That is a very good question. At least it is save with Elixir Github Action workflows.
I can't say for sure how tolerant Erlang is with this behavior.

Reply:
Yes it seem so that compilation is quick enough. So the benefit of caching whole `_build` is negligible.

Only dialyzer is slow and benefits a lot from caching the `plt`.
I would suggest to write a rule that contains the plt within `_build/dialyzer` only.

`.cache/rebar3` seems to be fine - not bringing much benefit - but not doing any harm either.

Reply:
I think `.cache/rebar3` is actually nice because it holds hex.pm cache (eg ranch) as well as dialyzer .plt for OTP standard library

Reply:
> I would suggest to write a rule that contains the plt within _build/dialyzer only.

Sounds like a good idea

Reply:
Done. I've narrowed done the cache specification.
- Hex packages
- Dialyzer PLTs

For now not caching any build files.
It would save some little seconds, but it's not worth it now until we have more experience with it.
(rebar3 is currently still determing recompilation by timestamps)

For the Dialyzer PLTs cache I would have preferred to have the full OTP version in the key.
But as PLTs are not that big of size and restoring cache is blazing fast I could probably forget about it.